contrib/natss/pkg/dispatcher/dispatcher/dispatcher.go

```diff
@@ -147,17 +181,35 @@ func (s *SubscriptionsSupervisor) connectWithRetry(stopCh <-chan struct{}) {
 	}
 }
 
+func (s *SubscriptionsSupervisor) signalReconcile(stopCh <-chan struct{}) {
+	s.subscriptionsMux.Lock()
+	defer s.subscriptionsMux.Unlock()
+
+	// Clear subscriptions to force resubscribe
+	s.subscriptions = make(map[provisioners.ChannelReference]map[subscriptionReference]*stan.Subscription)
+
+	// Capture + clear reonciliation
+	reconcilers := s.reconcilers
+	s.reconcilers = make(map[provisioners.ChannelReference]func(<-chan struct{}))
+
+	// Run in a goroutine since signalling can block
+	go func() {
+		for _, reconciler := range reconcilers {
+			reconciler(stopCh)
+		}
+	}()
+}
+
 // Connect is called for initial connection as well as after every disconnect
 func (s *SubscriptionsSupervisor) Connect(stopCh <-chan struct{}) {
 	for {
 		select {
 		case <-s.connect:
 			s.natssConnMux.Lock()
 			currentConnProgress := s.natssConnInProgress
-			s.natssConnMux.Unlock()
-			if !currentConnProgress {
-				// Case for lost connectivity, setting InProgress to true to prevent recursion
-				s.natssConnMux.Lock()
+			if currentConnProgress {
```

Comment:
This change fixes a race condition. At a high level this should be using an atomic check and set pattern (or at least an unlocked check -> lock -> recheck -> set if we're really worried about lock contention), but it was not before so `connectWithRetry` below could be called by multiple routines at the same time.